File: backend/app/search/verification.py
"""Live-price (or indicative-fallback) verification of the top candidate
groups -- spec §26 "LIVE FARE SEARCH" stage / §27 step 3. Hard
constraints (budget, stops, layover min/max) are enforced locally here
rather than trusted to any provider param, per spec intent.

Groups (see pruning.py) are verified whole: cheapest-group-first, and a
group is only verified if ALL of its origin variants fit in the
remaining live-call budget. This keeps a primary-origin result and its
alt-origin siblings verified together, which ranking.py's nearby-airport
filter depends on to compute a net saving.
"""

from __future__ import annotations

import asyncio
import datetime as dt
from dataclasses import dataclass

from sqlalchemy.ext.asyncio import AsyncSession

from app.config import Settings
from app.db import session_scope
from app.providers.base import FareOption, FareProvider, FareQuery
from app.search.pruning import Candidate, CandidateGroup
from app.search.spaces import SearchSpace
from app.services import fares as fare_service
from app.services import indicative

VERIFY_CONCURRENCY = 4


@dataclass
class VerifiedItinerary:
    origin: str
    destination: str
    depart_date: dt.date
    return_date: dt.date
    trip_length: int
    option: FareOption
    verified: bool  # False when we fell back to an indicative estimate (budget exhausted)


def passes_hard_constraints(option: FareOption, space: SearchSpace) -> bool:
    """Not underscore-prefixed: also called directly by
    search/pipeline.py's run_multi_city_search, which has no CandidateGroup
    to verify against and so never goes through verify_top/verify_one.
    """
    if option.slices:
        # Manual multi-city: validate each flown slice independently. The
        # deliberate gap BETWEEN slices (a multi-day city stop) is not a
        # layover and must never be checked against min/max_normal_minutes
        # -- that window means "is this a comfortable flight connection",
        # not "is this a reasonable amount of time to spend in a city".
        for sl in option.slices:
            if sl.stops > space.max_stops:
                return False
            for _, minutes in sl.layovers:
                if not (space.min_normal_minutes <= minutes <= space.max_normal_minutes):
                    return False
    else:
        if option.stops > space.max_stops:
            return False
        for _, minutes in option.layovers:
            if not (space.min_normal_minutes <= minutes <= space.max_normal_minutes):
                return False
    # Budget is interpreted per person (MVP 1 assumption; see plan risks --
    # adults=1 is the supported default, so this is exact in the common case).
    if option.price > space.max_total:
        return False
    return True


def _select_candidates_to_verify(groups: list[CandidateGroup], call_budget: int) -> list[Candidate]:
    selected_groups: list[CandidateGroup] = []
    used = 0
    for group in groups:
        if used + len(group.candidates) > call_budget:
            continue  # doesn't fit as a whole group -- try smaller ones rather than splitting it
        selected_groups.append(group)
        used += len(group.candidates)
        if used >= call_budget:
            break

    if selected_groups:
        return [c for g in selected_groups for c in g.candidates]

    if not groups:
        return []

    # Budget too tight for even the cheapest whole group -- verify as much
    # of it as fits rather than verifying nothing. This sacrifices the
    # pairing guarantee for this one group, but only under a budget
    # tighter than any real MVP 1 default.
    cheapest = groups[0]
    return sorted(cheapest.candidates, key=lambda c: c.estimated_price)[:call_budget]


async def verify_top(
    session: AsyncSession,
    space: SearchSpace,
    groups: list[CandidateGroup],
    settings: Settings,
    *,
    provider: FareProvider | None = None,
) -> tuple[list[VerifiedItinerary], bool]:
    """Returns (verified_itineraries, degraded). degraded=True if any
    candidate had to fall back to an indicative estimate because the
    live budget was exhausted mid-search.

    `provider` is the caller's own per-user FareProvider (api/deps.py) --
    passed through untouched to fare_service.search_cached, which is what
    actually skips the site-wide budget guard for it. None only for
    site-wide-default callers (currently none in the real request path;
    kept for tests / any future site-wide fallback).

    The `session` parameter is intentionally unused by the concurrent path
    below -- an AsyncSession is not safe to share across coroutines running
    concurrently under asyncio.gather, so each verify_one() opens its own
    short-lived session instead (see app/db.py's session_scope). `session`
    stays in the signature for consistency with the other pipeline stages,
    which all run sequentially and do share one session safely.
    """
    call_budget = min(settings.verify_top_n, settings.per_search_live_cap)
    to_verify = _select_candidates_to_verify(groups, call_budget)

    semaphore = asyncio.Semaphore(VERIFY_CONCURRENCY)
    any_degraded = False

    async def verify_one(candidate: Candidate) -> VerifiedItinerary | None:
        nonlocal any_degraded
        query = FareQuery(
            origin=candidate.origin,
            destination=candidate.destination,
            depart_date=candidate.depart_date,
            return_date=candidate.return_date,
            adults=space.adults,
            currency=space.currency,
            max_stops=space.max_stops,
            trip_type=space.trip_type,
        )
        async with semaphore, session_scope() as local_session:
            options, degraded = await fare_service.search_cached(
                local_session, query, kind="verified", provider=provider
            )

            if degraded:
                any_degraded = True
                price, _source = await indicative.estimate(
                    local_session, candidate.origin, candidate.destination, candidate.depart_date, candidate.return_date,
                    trip_type=space.trip_type,
                )
                if price > space.max_total:
                    return None
                fallback = FareOption(
                    price=price,
                    currency=space.currency,
                    outbound_legs=(),
                    layovers=(),
                    total_duration_min=0,
                    stops=0,
                    carriers=(),
                    inbound_detail="indicative",
                )
                return VerifiedItinerary(
                    candidate.origin, candidate.destination, candidate.depart_date, candidate.return_date,
                    candidate.trip_length, fallback, verified=False,
                )

        passing = [o for o in options if passes_hard_constraints(o, space)]
        if not passing:
            return None
        best = min(passing, key=lambda o: o.price)
        return VerifiedItinerary(
            candidate.origin, candidate.destination, candidate.depart_date, candidate.return_date,
            candidate.trip_length, best, verified=True,
        )

    results = await asyncio.gather(*(verify_one(c) for c in to_verify))
    verified = [r for r in results if r is not None]
    return verified, any_degraded

File: frontend/src/lib/engine/verification.ts
// Live-price (or indicative-fallback) verification of the top candidates
// -- cheapest/most-promising first, up to the live-call budget.
//
// "Degraded" here means a live call FAILED (network, SerpApi error,
// quota). The candidate then falls back to this user's own past
// observation for the same route/dates/party if one exists (a REAL past
// price, not a guess) and is flagged verified=false; with no such
// observation there is nothing honest to show, so the candidate is
// dropped rather than displaying a fabricated number.

import { FARE_CACHE_TTL_EMPTY_MS, FARE_CACHE_TTL_VERIFIED_MS, PER_SEARCH_LIVE_CAP, VERIFY_CONCURRENCY, VERIFY_TOP_N } from "./constants";
import { addDaysIso } from "./dates";
import { estimate, partyKey } from "./indicative";
import { addObservations, cacheGet, cacheSet, recordLiveCall } from "./observations";
import { ProxyError, SerpApiError } from "./serpapi";
import type { Candidate, EngineProvider, FareOption, FareQuery, SearchSpace, VerifiedItinerary } from "./types";

export function passesHardConstraints(option: FareOption, space: SearchSpace): boolean {
  if (option.slices.length > 0) {
    // Manual multi-city: validate each flown slice independently. The
    // deliberate gap BETWEEN slices (a multi-day city stop) is not a
    // layover and must never be checked against the connection-comfort
    // window -- that window means "is this a comfortable flight
    // connection", not "is this a reasonable time to spend in a city".
    for (const sl of option.slices) {
      if (sl.stops > space.maxStops) return false;
      for (const [, minutes] of sl.layovers) {
        if (minutes < space.minNormalMinutes || minutes > space.maxNormalMinutes) return false;
      }
    }
  } else {
    if (option.stops > space.maxStops) return false;
    for (const [, minutes] of option.layovers) {
      if (minutes < space.minNormalMinutes || minutes > space.maxNormalMinutes) return false;
    }
  }
  // SerpApi's price is always for the whole party -- max_total is asked
  // for, and must be compared, the same way.
  if (option.price > space.maxTotal) return false;
  return true;
}

export function cacheKey(providerName: string, query: FareQuery): string {
  return `fare:v2:${providerName}:${query.origin}:${query.destination}:${query.departDate}:${query.returnDate}:${query.currency}:${query.maxStops}:${query.tripType}:${partyKey(query.passengers)}:${query.travelClass}`;
}

async function fetchOptions(
  provider: EngineProvider,
  query: FareQuery,
  signal?: AbortSignal,
): Promise<{ options: FareOption[]; degraded: boolean }> {
  const key = cacheKey(provider.name, query);
  const cached = cacheGet(key);
  if (cached !== null) return { options: cached, degraded: false };

  let options: FareOption[];
  try {
    options =
      query.tripType === "one_way"
        ? await provider.searchOneWay(query, signal)
        : await provider.searchRoundTrip(query, signal);
  } catch (err) {
    if (signal?.aborted) throw err;
    if (err instanceof SerpApiError || err instanceof ProxyError) {
      return { options: [], degraded: true };
    }
    throw err;
  }

  if (provider.name !== "mock") recordLiveCall();
  cacheSet(key, options, options.length > 0 ? FARE_CACHE_TTL_VERIFIED_MS : FARE_CACHE_TTL_EMPTY_MS);
  if (options.length > 0) {
    const now = Date.now();
    addObservations(
      options.map((o) => ({
        origin: query.origin,
        destination: query.destination,
        departure_date: query.departDate,
        return_date: query.returnDate,
        trip_type: query.tripType,
        fare: o.price,
        observed_at: now,
        party_key: partyKey(query.passengers),
      })),
    );
  }
  return { options, degraded: false };
}

export async function verifyTop(
  space: SearchSpace,
  candidates: Candidate[],
  provider: EngineProvider,
  signal?: AbortSignal,
): Promise<{ verified: VerifiedItinerary[]; degraded: boolean }> {
  const callBudget = Math.min(VERIFY_TOP_N, PER_SEARCH_LIVE_CAP);
  // pruning.ts already sorted candidates signal-first/cheapest-first --
  // just take the top of the live-call budget.
  const toVerify = candidates.slice(0, callBudget);

  let anyDegraded = false;

  async function verifyOne(candidate: Candidate): Promise<VerifiedItinerary | null> {
    const query: FareQuery = {
      origin: space.originGroup,
      destination: candidate.destination.joined,
      departDate: candidate.departDate,
      returnDate: candidateReturnDate(candidate),
      passengers: space.passengers,
      travelClass: space.travelClass,
      currency: space.currency,
      maxStops: space.maxStops,
      tripType: space.tripType,
    };

    const { options, degraded } = await fetchOptions(provider, query, signal);

    if (degraded) {
      anyDegraded = true;
      const [price] = estimate(query.origin, query.destination, query.departDate, query.returnDate, space.tripType, space.passengers);
      if (price === null || price > space.maxTotal) return null;
      const fallback: FareOption = {
        price,
        currency: space.currency,
        outbound_legs: [],
        layovers: [],
        total_duration_min: 0,
        stops: 0,
        carriers: [],
        inbound_detail: "indicative",
        slices: [],
      };
      return {
        destination: candidate.destination,
        departDate: candidate.departDate,
        returnDate: query.returnDate,
        tripLength: candidate.tripLength,
        option: fallback,
        verified: false,
      };
    }

    const passing = options.filter((o) => passesHardConstraints(o, space));
    if (passing.length === 0) return null;
    const best = passing.reduce((a, b) => (b.price < a.price ? b : a));
    return {
      destination: candidate.destination,
      departDate: candidate.departDate,
      returnDate: query.returnDate,
      tripLength: candidate.tripLength,
      option: best,
      verified: true,
    };
  }

  // Concurrency-limited pool (the original backend used an asyncio
  // semaphore of 4).
  const results: (VerifiedItinerary | null)[] = new Array(toVerify.length).fill(null);
  let next = 0;
  async function workerLoop(): Promise<void> {
    while (next < toVerify.length) {
      if (signal?.aborted) throw new DOMException("Aborted", "AbortError");
      const index = next++;
      results[index] = await verifyOne(toVerify[index]);
    }
  }
  await Promise.all(Array.from({ length: Math.min(VERIFY_CONCURRENCY, toVerify.length) }, workerLoop));

  return { verified: results.filter((r): r is VerifiedItinerary => r !== null), degraded: anyDegraded };
}

function candidateReturnDate(candidate: Candidate): string {
  // return_date is derived, not stored: depart + tripLength days.
  // tripLength 0 (the one-way sentinel) naturally yields returnDate === departDate.
  return addDaysIso(candidate.departDate, candidate.tripLength);
}
